mozregression/fetch_configs.py

```diff
@@ -91,24 +91,39 @@ def get_build_regex(name, os, bits, processor, psuffix='', with_ext=True):
         return regex
 
 
-def _extract_build_type(build_type):
-    """Internal function to return a list from a build type string"""
-    return [t.strip() for t in build_type.split(',')]
-
-
 class CommonConfig(object):
     """
     Define the configuration for both nightly and inbound fetching.
     """
     BUILD_TYPES = ('opt',)  # only opt allowed by default
+    BUILD_TYPE_FALLBACKS = {}
     app_name = None
 
     def __init__(self, os, bits, processor):
         self.os = os
         self.bits = bits
         self.processor = processor
-        self.build_type = 'opt'
         self.repo = None
+        self.set_build_type('opt')
+        self._used_build_index = 0
+
+    @property
+    def build_type(self):
+        """
+        Returns the currently selected build type, which can change if there
+        are fallbacks specified.
+        """
+        return self.build_types[self._used_build_index]
+
+    def inc_used_build(self):
+        """
+        Increments the index into the build_types indicating the currently
+        selected build type.
+        """
+        self._used_build_index = (
+            # Need to be careful not to overflow the list
+            (self._used_build_index + 1) % len(self.build_types)
```

Comment:
Should we throw an exception if we overflow? I'm pretty sure that should never happen.

Reply:
Well if it does ever overflow and build_type is called we'll get `IndexError: list index out of range` anyway (which is what made me realise I need to add the mod operation to prevent it), so I think we should be fine.

Reply:
Hmm, does this mean we might have an infinite loop? Seems like it would be better to bail out, wouldn't it?

Reply:
Well nothing should be iterating using the index anyway (since I realised I could just make `tk_inbound_routes` a generator and track the used build internally), they should just iterate `tk_inbound_routes` to get them all, and `tk_inbound_routes` doesn't actually use it, just iterates the build_types and set this.  I suppose we could have it just stop incrementing once it reaches the end.  I think the original reason I had it loop was so it would end-up "reset", and report it as looking for the original requested build rather than fallbacks on total failure, but I guess build_type isn't actually used in that instance, and it is reporting the last fallback route as the failure anyway.

Reply:
Ok, sounds reasonable! I feel like we could simplify this area of the code a bunch, but that might require further refactoring. For now this looks like a good step forward.